docs/configuration.md
# Configuration

Application configuration is handled using [pydantic-settings](https://docs.pydantic.dev/latest/concepts/pydantic_settings/).

The bot is configured to look for environment variebles from the sys environment and a `.env` file (dotenv) in that order.

All other configuration options are handled by creating a `config.yaml` file. Almost all settings have default values, so you can run the application without creating a `config.yaml` file, but if you wish to enable integrations or change various settings, you'll want to create one.

Please review the [integrations](integrations.md) documentation for additional information on enabling and configuring integrations.

## Configurable Parameters

### API

Additional API routes are disabled by default. This flag only needs to be set to `true` if using the web interface.

Note that the API exposes a health check route.

!!! note

    The web interface is optional and is not required to use the application.

```yaml
api:
  enabled: true
```

With the API enabled, routes will be made available via `/api/v1`.

!!! warning

    API routes are only meant to serve the web interface. As such, they are secured using JWT and will not work without first running through the setup guide for the [web interface](ui.md).

!!! warning

    The default username and password for the web interface are set to `admin@example.com`:`changethis`. It is **strongly** recommended you change these variables before exposing the API.

    Set the environment variables `FIRST_SUPERUSER` and `FIRST_SUPERUSER_PASSWORD` to unique values before running the application the first time.

```python
# incidentbot/configuration/settings.py
FIRST_SUPERUSER: str = "admin@example.com"
FIRST_SUPERUSER_PASSWORD: str = "changethis"
```

#### Documentation Endpoints

All endpoints related to documentation for the API are **disabled** by default. You don't have to provide these values in the settings configuration if you don't wish to override them.

To enable them, use the following settings:

```yaml
# Set any of these to true to enable them.
api:
  enable_docs_endpoint: false # /docs
  enable_openapi_endpoint: false # /openapi.json
  enable_redoc_endpoint: false # /redoc
```

It is recommended to keep these disabled when the API is exposed to the general public.

### Digest Channel

The digest channel is where updates are sent regarding all incidents managed by the bot. The channel is `#incidents` by default.

To change it, set the following value in `config.yaml`:

```yaml
# This is the default value.
digest_channel: incidents
```

### Pinned Images

Pinning images to incident timelines is enabled by default.

To disable this behavior, set the following value in `config.yaml`:

```yaml
enable_pinned_images: false
```

!!! note

    You may not be able to pin images on a free/trial plan. The API method that it requires, `files.sharedPublicURL`, typically won't work.

    That being said, it has worked fine in the past despite the error that will show in the logs.

### Initial Communications Reminders

The bot will watch all incidents that are created and will send notifications to the channel to remind participants to send out updates and claim roles if they remain unclaimed.

To change the amount of time the bot waits before sending these notifications, set the following values in `config.yaml`:

```yaml
# These are the default values.
initial_comms_reminder_minutes: 30
initial_role_watcher_minutes: 10
```

### Links

foo do these even get used?

### Maintenance Windows

The bot has a feature that will send out notifications for scheduled maintenance. This feature is disabled by default.

To enabled this feature, set the following value in `config.yaml`:

```yaml
maintenance_windows:
  components:
    - API
    - Auth
    - Databases
    - Website
```

By setting a list of components that will show up as components related to a maintenance window, the feature will be enabled.

By default, statuses for maintenance windows are `Scheduled`, `In Progress`, `Complete`. If you wish to override these values, you can add the following block to the settings:


```yaml
maintenance_windows:
  statuses:
    - Scheduled
    - In Progress
    - Complete
```

Note that statuses should be stated in order of start to finish. The first status should be the initial status, etc.

### Options

There is an `options` section that holds several settings.

To change any of these values, set the following values in `config.yaml`:

```yaml
options:
  additional_welcome_messages:
    # Any additional messages to add to a new incident when it is opened.
    - message: "Welcome to the incident. Please be sure to..."
      pin: true # false by default
  auto_invite_groups:
    - name: str
      # Only if PagerDuty integration is enabled
      # Create a page when these groups are invited
      pagerduty_escalation_policy: some-policy
      pagerduty_escalation_priority: high
      severities: sev1,sev2
      # severities: all
  # This is the default value. This is what all incident channels are prefixed with.
  channel_name_prefix: inc
  # There is no default value for this. If one is set, this URL will be used for the meeting advertised with each incident.
  # Will not be used if using automatic Zoom meeting generation.
  meeting_link: None
  # If true, pin the meeting link to the incident channel upon creation.
  pin_meeting_link_to_channel: false
  # This limits the amount of incidents shown on the summary on the app home page.
  # It is not recommended to raise this value very high due to Slack limitations on how many blocks can appear in a message.
  # This is the default value.
  show_most_recent_incidents_app_home_limit: 5
  # Application timezone.
  # This is the default value.
  timezone: UTC
```

### Platform

Right now, the only valid value for the `platform` field is `slack`.

### Roles

You can use the default incident management roles or define your own.

Roles are configured using the following format. The "commander," or "lead" role, should be marked with `is_lead`.

!!! note

    You do not need to provide the default values displayed below if you wish to use the default values. This is provided as an example of how to configure custom values.

```yaml
roles:
  incident_commander:
    description: "The Incident Commander is the decision maker during a major incident, delegating tasks and listening to input from subject matter experts in order to bring the incident to resolution. They become the highest ranking individual on any major incident call, regardless of their day-to-day rank. Their decisions made as commander are final.\n\nYour job as an Incident Commander is to listen to the call and to watch the incident Slack room in order to provide clear coordination, recruiting others to gather context and details. You should not be performing any actions or remediations, checking graphs, or investigating logs. Those tasks should be delegated.\n\nAn IC should also be considering next steps and backup plans at every opportunity, in an effort to avoid getting stuck without any clear options to proceed and to keep things moving towards resolution.\n\nMore information: https://response.pagerduty.com/training/incident_commander/"
    is_lead: true
  scribe:
    description: "The purpose of the Scribe is to maintain a timeline of key events during an incident, documenting actions, and keeping track of any follow-up items that will need to be addressed.\n\nMore information: https://response.pagerduty.com/training/scribe/"
  subject_matter_expert:
    description: "A Subject Matter Expert (SME) is a domain expert or designated owner of a component or service that is part of the software stack. These are critical members of the incident response process that play pivotal roles in identifying and resolving individual components of impacted ecosystems.\n\nMore information: https://response.pagerduty.com/training/subject_matter_expert/"
  communications_liaison:
    description: "The purpose of the Communications Liaison is to be the primary individual in charge of notifying our customers of the current conditions, and informing the Incident Commander of any relevant feedback from customers as the incident progresses.\n\nIt's important for the rest of the command staff to be able to focus on the problem at hand, rather than worrying about crafting messages to customers.\nYour job as Communications Liaison is to listen to the call, watch the incident Slack room, and track incoming customer support requests, keeping track of what's going on and how far the incident is progressing (still investigating vs close to resolution).\n\nThe Incident Commander will instruct you to notify customers of the incident and keep them updated at various points throughout the call. You will be required to craft the message, gain approval from the IC, and then disseminate that message to customers.\n\nMore information: https://response.pagerduty.com/training/customer_liaison/"
```

### Severities

You can use the default incident management severities or define your own.

Severities are configured using the following format.

!!! note

    You do not need to provide the default values displayed below if you wish to use the default values. This is provided as an example of how to configure custom values.

```yaml
severities:
  sev1: "This signifies a critical production scenario that impacts most or all users with a major impact on SLAs. This is an all-hands-on-deck scenario that requires swift action to restore operation. Customers must be notified."
  sev2: "This signifies a significant production degradation scenario impacting a large portion of users."
  sev3: "This signifies a minor production scenario that may or may not result in degradation. This situation is worth coordination to resolve quickly but does not indicate a critical loss of service for users."
  sev4: "This signifies an ongoing investigation. This incident has not been promoted to SEV3 yet, indicating there may be little to no impact, but the situation warrants a closer look. This is diagnostic in nature. This is the default setting for a new incident."
```

### Statuses

You can use the default incident management statuses or define your own.

Statuses are configured using the following format.

!!! note

    You do not need to provide the default values displayed below if you wish to use the default values. This is provided as an example of how to configure custom values.

```yaml
statuses:
  investigating:
    initial: true
  identified: {}
  monitoring: {}
  resolved:
    final: true
```

It is important to mark a status as `initial` and a status as `final` either way. This tells the bot the starting and finishing value for incident statuses.

### Slash Command

The bot's default slash command is `/incidentbot`. If you wish to override it, you ca

To change this value, set the following value in `config.yaml`:

```yaml
# This is the default value.
root_slash_command: '/incidentbot'
```

!!! note

    If you change this value, be sure to update the manifest to match.

### Updates

When using the "provide incident update" functionality, default behavior is to add any updates for an incident as comments under the original thread in the digest channel. If you wish to send all updates to the digest channel directly instead, set the following value in `config.yaml`:

```yaml
options:
  updates_in_threads: true
```


docs/changelog.md
# Changelog

## v2.0.14 (2024-10-10)

- Builds an arm64 version of the util image.
- Adds tests for image suffixes to chart.
- Minor version bumps for Poetry-managed packages.

## v2.0.13 (2024-10-05)

- Fix bug with maintenance window management.

## v2.0.12 (2024-10-05)

- Fix bugs with Statuspage integration.
- This release requires a database migration - please plan accordingly. This is handled by the chart if using its init features.

## v2.0.9 (2024-09-30)

- Allow disabling the API docs endpoint.
- Updates to docs for web interface.

## v2.0.8 (2024-09-30)

- Corrections in documentation for web interface usage.
- Add some packages required for docs development.

## v2.0.7 (2024-09-30)

- Fixes bug where `postmortem_link` variable didn't work if integration was disabled.
- Fixes potential bugs where `this` command prompts resulted in errors due to responses from Slack.
- Bumps a few packages.

## v2.0.6 (2024-09-26)

- Fixes bug where PagerDuty information would not store in the database properly.
- Fixes bug where trying to create more than one PagerDuty incident per bot incident failed.
- Fixes bug where matching a bot incident when trying to create a PagerDuty incident failed.
- Fixes bug where creating a PagerDuty incident failed because the column that used to store incidents has been removed.
- PagerDuty incidents now create records properly.
- The message when a page is issue now contains a link to the incident it creates.

## v2.0.5 (2024-09-24)

- Fixes bug where `postmortem_link` was no longer a valid argument for generating a digest message update.
- Adds link to postmortem in a digest message if one is passed in when the method is called.
- Adds retry logic to all Slack API requests to avoid failures based on data not populating.
- Fixes bug where a user's real name being unavailable for parsing would cause other things to fail.

## v2.0.4 (2024-09-23)

- Fixes bug where the incident list functionality did not work via command or within app home in Slack.
- Fixes incorrect content in some Helm templates.
- Adds value to Helm chart to set database password if using built-in database.
- Fixes bug in Helm chart where the `ConfigMap` data to populate `config.yaml` did not mount anything, resulting in the app always using default values.
- Fixes some spelling mistakes in the documentation.
- Adds clarity to documentation around database migrations.
- Fixes bug where the `enterprise` value in Slack responses was assumed to always be `str`, but can actually be `dict`.
- Cleans up some minor errors around the pager modal.

## v2.0.3 (2024-09-20)

- Adds unit testing to Helm chart.
- Adds better readme for Helm chart.
- Minor chart cleanup and fixes.

## v2.0.1 (2024-09-20)

- Add init containers to Helm chart to handle database migrations.
- Add Dockerfile to build utility image for use with migrations and other tasks.
- Clarify in documentation any steps regarding database migrations.
- Add jobs to CircleCI build to test util and docs builds and build util image.
- Move all Dockerfiles to root and fix references.
- Update .gitignore file.
- Fix versioning script.

## v2.0.0 (2024-09-19)

- Poetry has been adopted for Python package management, replacing the legacy pattern.
- Pydantic is now used for both type-safety and convenience as well as configuration management.
- SQLModel is now used to manage database models, extending the usefulness of Pydantic.
- The database schema has changed entirely. It is recommended to start fresh if you have an existing deployment.
- FastAPI has replaced Flask as the API framework, further extending the usefulness of Pydantic.
- Configuration has been flattened and simplified. There is now no hard requirement to provide a `config.yaml` file if accepting all defaults.
- Statuses are now fully customizable and there are no hard dependencies on specific ones. You have the ability to decide initial and final statuses.
- There is now a feature to create and manage maintenance windows to advertise scheduled maintenance. This feature is new and will have more functionality added over time.
- There is now an option to create an additional "comms" channel alongside incident channels to separate critical communications. This is entirely optional.
- All message formatting has been rebuilt to be cleaner and easier to look at.
- Reacji to pin content is now customizable.
- Reacting with reacji to messages to open incidents has been removed.
- Migrations are now required for all database operations. A base Alembic migration is provided and must be run prior to first startup.
- All use of shortcuts has been removed and replaced with slash commands. No more searching for specific commands.
- All interaction is handled using ephemeral messages to reduce clutter in public channels.
- Timestamp issues have been fixed by removing storing string timestamps in the database as strings and handling this via native database time columns.
- Introduces patterns to adopt more providers than Slack in future releases.
- Logs can now be pretty printed or JSON.
- There are new jobs for communications reminders and role watcher. These jobs will send messages to an incident channel to remind participants to send out regular updates and to claim unclaimed roles. Either can be disabled or dismissed.
- Postmortem generation now requires a Confluence template. The bot will no longer generate an entire page layout for you. There are template strings that can be placed in a template to inject values from the incident, leaving the rest up to you.
- The UI has been replaced with a new framework largely inspired by FastAPI's fullstack example.


incidentbot/configuration/settings.py
import os
import secrets

from pydantic import (
    AnyUrl,
    BaseModel,
    BeforeValidator,
    computed_field,
    model_validator,
    TypeAdapter,
)
from pydantic_settings import (
    BaseSettings,
    PydanticBaseSettingsSource,
    SettingsConfigDict,
    YamlConfigSettingsSource,
)
from typing import Annotated, Any, Tuple, Type
from typing_extensions import Self

__version__ = "v2.1.3"

opsgenie_logo_url = "https://i.imgur.com/NjiEBCu.png"
pagerduty_logo_url = "https://i.imgur.com/IVvdFCV.png"
statuspage_logo_url = "https://i.imgur.com/v4xmF6u.png"


def parse_cors(v: Any) -> list[str] | str:
    if isinstance(v, str) and not v.startswith("["):
        return [i.strip() for i in v.split(",")]
    elif isinstance(v, list | str):
        return v

    raise ValueError(v)


"""
API
"""


class API(BaseModel):
    enabled: bool | None = False
    enable_docs_endpoint: bool | None = False
    enable_openapi_endpoint: bool | None = False
    enable_redoc_endpoint: bool | None = False
    v1_str: str | None = "/api/v1"


"""
Jobs
"""


class ScrapeForAgingIncidentsJob(BaseModel):
    """
    Model for the jobs scrape_for_aging_incidents_job field
    """

    enabled: bool = True
    ignore_statuses: list = []


class Jobs(BaseModel):
    """
    Model for the jobs field
    """

    scrape_for_aging_incidents: ScrapeForAgingIncidentsJob


"""
Options
"""


class AdditionalWelcomeMessage(BaseModel):
    """
    Model for messages that will be added to the beginning of all incidents
    """

    message: str
    pin: bool | None = False


class GroupAutoInvite(BaseModel):
    """
    Model for groups that should be auto invited to incidents
    """

    name: str
    pagerduty_escalation_policy: str | None = None
    pagerduty_escalation_priority: str | None = "low"
    severities: str | None = "all"


class Options(BaseModel):
    """
    Model for the options field
    """

    additional_welcome_messages: list[AdditionalWelcomeMessage] | None = None
    auto_invite_groups: list[GroupAutoInvite] | None = None
    channel_name_prefix: str | None = "inc"
    meeting_link: str | None = None
    pin_meeting_link_to_channel: bool = False
    skip_logs_for_user_agent: list[str] | None = None
    show_most_recent_incidents_app_home_limit: int = 5
    slack_items_pagination_per_page: int = 5
    timezone: str = "UTC"
    updates_in_threads: bool | None = False


"""
Integrations
"""


class ConfluenceIntegration(BaseModel):
    """
    Model for the confluence field
    """

    auto_create_postmortem: bool | None = False
    enabled: bool = False
    parent: str
    space: str
    template_id: int


class JiraIntegration(BaseModel):
    """
    Model for the jira field
    """

    auto_create_issue: bool = False
    auto_create_issue_type: str | None = None
    enabled: bool = False
    issue_types: list[str]
    labels: list[str] | None = None
    priorities: list[str] | None = None
    project: str
    status_mapping: list[dict[str, str]]


class OpsgenieIntegration(BaseModel):
    """
    Model for the opsgenie field
    """

    enabled: bool = False
    team: str


class StatuspageIntegrationPermissions(BaseModel):
    """
    Model for the statuspage permissions field
    """

    groups: list[str] | None = None


class StatuspageIntegration(BaseModel):
    """
    Model for the statuspage field
    """

    enabled: bool = False
    permissions: StatuspageIntegrationPermissions | None = None
    url: str


class AtlassianIntegration(BaseModel):
    """
    Model for the atlaassian field
    """

    confluence: ConfluenceIntegration | None = None
    jira: JiraIntegration | None = None
    opsgenie: OpsgenieIntegration | None = None
    statuspage: StatuspageIntegration | None = None


class PagerDutyIntegration(BaseModel):
    """
    Model for the pagerduty field
    """

    enabled: bool = False


class ZoomIntegration(BaseModel):
    """
    Model for the zoom field
    """

    auto_creating_meeting: bool
    enabled: bool = False


class Integrations(BaseModel):
    """
    Model for the integrations field
    """

    atlassian: AtlassianIntegration | None = None
    pagerduty: PagerDutyIntegration | None = None
    zoom: ZoomIntegration | None = None


"""
Root
"""


class Link(BaseModel):
    """
    Model for the links field
    """

    title: str
    url: str


class MaintenanceWindows(BaseModel):
    """
    Model for the maintenance_windows field
    """

    components: list[str]
    statuses: list[str] | None = ["Scheduled", "In Progress", "Complete"]


class RoleDefinition(BaseModel):
    """
    Model for defining roles for incident participants
    """

    description: str
    is_lead: bool | None = False


class StatusDefinition(BaseModel):
    """
    Model for defining statuses for incidents
    """

    initial: bool | None = False
    final: bool | None = False


class Settings(BaseSettings):
    """
    Root settings model
    """

    """
    yaml
    """

    api: API | None = API()
    digest_channel: str = "incidents"
    emails_enabled: bool = False
    enable_pinned_images: bool = True
    icons: dict[str, dict[str, str]] = {
        "slack": {
            "channel": ":slack:",
            "components": ":jigsaw:",
            "description": ":mag_right:",
            "impact": ":chart_with_upwards_trend:",
            "maintenance": ":hammer_and_wrench:",
            "meeting": ":busts_in_silhouette:",
            "postmortem": ":book:",
            "role": ":bust_in_silhouette:",
            "status": ":fire_extinguisher:",
            "stopwatch": ":stopwatch:",
            "severity": ":rotating_light:",
            "task": ":ballot_box_with_check:",
            "update": ":incoming_envelope:",
        },
    }
    initial_comms_reminder_minutes: int = 30
    initial_role_watcher_minutes: int = 10
    integrations: Integrations | None = None
    jobs: Jobs | None = None
    links: list[Link] | None = None
    maintenance_windows: MaintenanceWindows | None = None
    options: Options | None = Options()
    pin_content_reacji: str = "pushpin"
    platform: str = "slack"
    roles: dict[str, RoleDefinition] = {
        "incident_commander": {
            "description": "The Incident Commander is the decision maker during a major incident, delegating tasks and listening to input from subject matter experts in order to bring the incident to resolution. They become the highest ranking individual on any major incident call, regardless of their day-to-day rank. Their decisions made as commander are final.\n\nYour job as an Incident Commander is to listen to the call and to watch the incident Slack room in order to provide clear coordination, recruiting others to gather context and details. You should not be performing any actions or remediations, checking graphs, or investigating logs. Those tasks should be delegated.\n\nAn IC should also be considering next steps and backup plans at every opportunity, in an effort to avoid getting stuck without any clear options to proceed and to keep things moving towards resolution.\n\nMore information: https://response.pagerduty.com/training/incident_commander/",
            "is_lead": True,
        },
        "scribe": {
            "description": "The purpose of the Scribe is to maintain a timeline of key events during an incident, documenting actions, and keeping track of any follow-up items that will need to be addressed.\n\nMore information: https://response.pagerduty.com/training/scribe/",
        },
        "subject_matter_expert": {
            "description": "A Subject Matter Expert (SME) is a domain expert or designated owner of a component or service that is part of the software stack. These are critical members of the incident response process that play pivotal roles in identifying and resolving individual components of impacted ecosystems.\n\nMore information: https://response.pagerduty.com/training/subject_matter_expert/",
        },
        "communications_liaison": {
            "description": "The purpose of the Communications Liaison is to be the primary individual in charge of notifying our customers of the current conditions, and informing the Incident Commander of any relevant feedback from customers as the incident progresses.\n\nIt's important for the rest of the command staff to be able to focus on the problem at hand, rather than worrying about crafting messages to customers.\nYour job as Communications Liaison is to listen to the call, watch the incident Slack room, and track incoming customer support requests, keeping track of what's going on and how far the incident is progressing (still investigating vs close to resolution).\n\nThe Incident Commander will instruct you to notify customers of the incident and keep them updated at various points throughout the call. You will be required to craft the message, gain approval from the IC, and then disseminate that message to customers.\n\nMore information: https://response.pagerduty.com/training/customer_liaison/"
        },
    }
    root_slash_command: str = "/incidentbot"
    severities: dict[str, str] = {
        "sev1": "This signifies a critical production scenario that impacts most or all users with a major impact on SLAs. This is an all-hands-on-deck scenario that requires swift action to restore operation. Customers must be notified.",
        "sev2": "This signifies a significant production degradation scenario impacting a large portion of users.",
        "sev3": "This signifies a minor production scenario that may or may not result in degradation. This situation is worth coordination to resolve quickly but does not indicate a critical loss of service for users.",
        "sev4": "This signifies an ongoing investigation. This incident has not been promoted to SEV3 yet, indicating there may be little to no impact, but the situation warrants a closer look. This is diagnostic in nature. This is the default setting for a new incident.",
    }
    statuses: dict[str, StatusDefinition] = {
        "investigating": {
            "initial": True,
        },
        "identified": {},
        "monitoring": {},
        "resolved": {
            "final": True,
        },
    }

    """
    .env
    """

    ACCESS_TOKEN_EXPIRE_MINUTES: int = 60 * 24 * 8
    SECRET_KEY: str = secrets.token_urlsafe(32)

    DOMAIN: str = "localhost"
    SMTP_TLS: bool = True
    SMTP_SSL: bool = False
    SMTP_PORT: int = 587
    SMTP_HOST: str | None = None
    SMTP_USER: str | None = None
    SMTP_PASSWORD: str | None = None

    EMAILS_FROM_EMAIL: str | None = None
    EMAILS_FROM_NAME: str | None = None
    EMAIL_RESET_TOKEN_EXPIRE_HOURS: int = 48

    FIRST_SUPERUSER: str = "admin@example.com"
    FIRST_SUPERUSER_PASSWORD: str = "changethis"

    BACKEND_CORS_ORIGINS: Annotated[
        list[AnyUrl] | str, BeforeValidator(parse_cors)
    ] = []

    POSTGRES_DB: str
    POSTGRES_HOST: str
    POSTGRES_PASSWORD: str
    POSTGRES_PORT: int
    POSTGRES_USER: str

    @computed_field  # type: ignore[prop-decorator]
    @property
    def DATABASE_URI(self) -> str:
        return f"postgresql://{self.POSTGRES_USER}:{self.POSTGRES_PASSWORD}@{self.POSTGRES_HOST}:{self.POSTGRES_PORT}/{self.POSTGRES_DB}"

    ATLASSIAN_API_URL: str | None = None
    ATLASSIAN_API_USERNAME: str | None = None
    ATLASSIAN_API_TOKEN: str | None = None
    ATLASSIAN_OPSGENIE_API_KEY: str | None = None
    ATLASSIAN_OPSGENIE_API_TEAM_INTEGRATION_KEY: str | None = None

    BETTERSTACK_UPTIME_API_TOKEN: str | None = None

    PAGERDUTY_API_TOKEN: str | None = None
    PAGERDUTY_API_USERNAME: str | None = None

    SLACK_APP_TOKEN: str | None = None
    SLACK_BOT_TOKEN: str | None = None
    SLACK_USER_TOKEN: str | None = None

    STATUSPAGE_API_KEY: str | None = None
    STATUSPAGE_PAGE_ID: str | None = None

    ZOOM_ACCOUNT_ID: str | None = None
    ZOOM_CLIENT_ID: str | None = None
    ZOOM_CLIENT_SECRET: str | None = None

    IS_MIGRATION: bool | None = False
    IS_TEST_ENVIRONMENT: bool | None = False

    LOG_LEVEL: str = "INFO"
    LOG_TYPE: str | None = None

    @computed_field  # type: ignore[prop-decorator]
    @property
    def server_host(self) -> str:
        # Use HTTPS for anything other than local development
        if self.ENVIRONMENT == "local":
            return f"http://{self.DOMAIN}"
        return f"https://{self.DOMAIN}"

    model_config = SettingsConfigDict(
        env_file=".env",
        env_ignore_empty=True,
        yaml_file=os.getenv("CONFIG_FILE_PATH", "config.yaml"),
    )

    def _check_required_var(self, var_name: str, value: str | None) -> None:
        if not value:
            message = f"The value of {var_name} cannot be empty."
            raise ValueError(message)

    def _check_required_integration_var(
        self, var_name: str, value: str | None, integration: str
    ) -> None:
        if not value:
            message = f"The value of {var_name} cannot be empty when enabling the {integration} integration."
            raise ValueError(message)

    @model_validator(mode="after")
    def _check_required_vars(self) -> Self:
        self._check_required_var("POSTGRES_DB", self.POSTGRES_DB)
        self._check_required_var("POSTGRES_HOST", self.POSTGRES_HOST)
        self._check_required_var("POSTGRES_PASSWORD", self.POSTGRES_PASSWORD)
        self._check_required_var("POSTGRES_PORT", self.POSTGRES_PORT)
        self._check_required_var("POSTGRES_USER", self.POSTGRES_USER)

        if not (
            TypeAdapter(bool).validate_python(self.IS_MIGRATION)
            or TypeAdapter(bool).validate_python(self.IS_TEST_ENVIRONMENT)
        ):
            self._check_required_var("SLACK_APP_TOKEN", self.SLACK_APP_TOKEN)
            self._check_required_var("SLACK_BOT_TOKEN", self.SLACK_BOT_TOKEN)
            self._check_required_var("SLACK_USER_TOKEN", self.SLACK_USER_TOKEN)

            if (
                self.integrations
                and self.integrations.atlassian
                and self.integrations.atlassian.confluence
                and self.integrations.atlassian.confluence.enabled
            ):
                self._check_required_integration_var(
                    "ATLASSIAN_API_URL", self.ATLASSIAN_API_URL, "Confluence"
                )
                self._check_required_integration_var(
                    "ATLASSIAN_API_USERNAME",
                    self.ATLASSIAN_API_USERNAME,
                    "Confluence",
                )
                self._check_required_integration_var(
                    "ATLASSIAN_API_TOKEN",
                    self.ATLASSIAN_API_TOKEN,
                    "Confluence",
                )

            if (
                self.integrations
                and self.integrations.atlassian
                and self.integrations.atlassian.jira
                and self.integrations.atlassian.jira.enabled
            ):
                self._check_required_integration_var(
                    "ATLASSIAN_API_URL", self.ATLASSIAN_API_URL, "Jira"
                )
                self._check_required_integration_var(
                    "ATLASSIAN_API_USERNAME",
                    self.ATLASSIAN_API_USERNAME,
                    "Jira",
                )
                self._check_required_integration_var(
                    "ATLASSIAN_API_TOKEN", self.ATLASSIAN_API_TOKEN, "Jira"
                )

            if (
                self.integrations
                and self.integrations.atlassian
                and self.integrations.atlassian.opsgenie
                and self.integrations.atlassian.opsgenie.enabled
            ):
                if self.integrations.atlassian.opsgenie.team:
                    self._check_required_integration_var(
                        "ATLASSIAN_OPSGENIE_API_TEAM_INTEGRATION_KEY",
                        self.ATLASSIAN_OPSGENIE_API_TEAM_INTEGRATION_KEY,
                        "Opsgenie",
                    )
                else:
                    self._check_required_integration_var(
                        "ATLASSIAN_OPSGENIE_API_KEY",
                        self.ATLASSIAN_OPSGENIE_API_KEY,
                        "Opsgenie",
                    )

            if (
                self.integrations
                and self.integrations.atlassian
                and self.integrations.atlassian.statuspage
                and self.integrations.atlassian.statuspage.enabled
            ):
                self._check_required_integration_var(
                    "STATUSPAGE_API_KEY", self.STATUSPAGE_API_KEY, "Statuspage"
                )
                self._check_required_integration_var(
                    "STATUSPAGE_PAGE_ID",
                    self.STATUSPAGE_PAGE_ID,
                    "Statuspage",
                )

            if (
                self.integrations
                and self.integrations.pagerduty
                and self.integrations.pagerduty.enabled
            ):
                self._check_required_integration_var(
                    "PAGERDUTY_API_USERNAME",
                    self.PAGERDUTY_API_USERNAME,
                    "PagerDuty",
                )
                self._check_required_integration_var(
                    "PAGERDUTY_API_TOKEN",
                    self.PAGERDUTY_API_TOKEN,
                    "PagerDuty",
                )

            if (
                self.integrations
                and self.integrations.zoom
                and self.integrations.zoom.enabled
            ):
                self._check_required_integration_var(
                    "ZOOM_ACCOUNT_ID", self.ZOOM_ACCOUNT_ID, "Zoom"
                )
                self._check_required_integration_var(
                    "ZOOM_CLIENT_ID", self.ZOOM_CLIENT_ID, "Zoom"
                )
                self._check_required_integration_var(
                    "ZOOM_CLIENT_SECRET", self.ZOOM_CLIENT_SECRET, "Zoom"
                )

        return self

    @classmethod
    def settings_customise_sources(
        cls,
        settings_cls: Type[BaseSettings],
        init_settings: PydanticBaseSettingsSource,
        env_settings: PydanticBaseSettingsSource,
        dotenv_settings: PydanticBaseSettingsSource,
        file_secret_settings: PydanticBaseSettingsSource,
    ) -> Tuple[PydanticBaseSettingsSource, ...]:
        return (
            env_settings,
            YamlConfigSettingsSource(settings_cls),
            dotenv_settings,
        )


settings = Settings()
